src/main/java/edu/greenblitz/tobyDetermined/subsystems/swerve/SwerveChassis.java

```diff
@@ -0,0 +1,459 @@
+		poseEstimator.update(getPigeonAngle(), getSwerveModulePositions());
+		Photonvision.getInstance().getUpdatedPoseEstimator().ifPresent((EstimatedRobotPose pose) -> poseEstimator.addVisionMeasurement(pose.estimatedPose.toPose2d(), pose.timestampSeconds));
+	}
+	
+	public void updatePoseEstimationLimeLight() {
+		if(MultiLimelight.getInstance().getFirstAvailableTarget().isPresent() || odometry.getPoseMeters().getTranslation().getDistance(getRobotPose().getTranslation()) < RobotMap.Odometry.MAX_DISTANCE_TO_FILTER_OUT) {
+			if(!robotHasObstacles()) {
+				poseEstimator.update(getGyroAngle(), getSwerveModulePositions());
+			}
+		}
+		if (doVision) {
+			if (MultiLimelight.getInstance().getAllEstimates().size() >= 1) {
+				twoApriltagsPresent = true;
+				for (Optional<Pair<Pose2d, Double>> target : MultiLimelight.getInstance().getAllEstimates()) {
+					target.ifPresent(this::addVisionMeasurement);
+				}
+			}
+		}
+	}
+	private boolean moduleHasObstacles(SwerveModule module) {
+		return module.getMotorOutputCurrent() > RobotMap.Swerve.SdsSwerve.FREE_CURRENT - currentTolerance && module.getMotorOutputCurrent() < currentTolerance + RobotMap.Swerve.SdsSwerve.FREE_CURRENT;
+	}
+	public boolean robotHasObstacles() {
+
+		boolean fl = moduleHasObstacles(frontLeft);
+		boolean fr = moduleHasObstacles(frontRight);
+		boolean bl = moduleHasObstacles(backLeft);
+		boolean br = moduleHasObstacles(backRight);
+
+		return ((fl && fr) || (bl && br) || (fl && bl) || (br && fr) || (fl && br) || (fr && bl));
+	}
+
+	public void updateOdometry() {
+		odometry.update(getGyroAngle(), getSwerveModulePositions());
+	}
+
+	public boolean getFlHasObstacles(){return moduleHasObstacles(frontLeft);}
+	public boolean getFrHasObstacles(){return moduleHasObstacles(frontRight);}
+	public boolean getBlHasObstacles(){return moduleHasObstacles(backLeft);}
+	public boolean getBrHasObstacles(){return moduleHasObstacles(backRight);}
+
+
+
+	private void addVisionMeasurement(Pair<Pose2d, Double> poseTimestampPair) {
+		Pose2d visionPose = poseTimestampPair.getFirst();
+		if (!(visionPose.getTranslation().getDistance(SwerveChassis.getInstance().getRobotPose().getTranslation()) > RobotMap.Vision.MIN_DISTANCE_TO_FILTER_OUT)) {
+			resetToVision();
+		}
+	}
+	public Pose2d getRobotPose() {
+		return poseEstimator.getEstimatedPosition();
+	}
+	
+	public void resetToVision() {
+		Optional<Pair<Pose2d, Double>> visionOutput = MultiLimelight.getInstance().getFirstAvailableTarget();
+		if(visionOutput.isPresent()) {
+			poseEstimator.setVisionMeasurementStdDevs(new MatBuilder<>(Nat.N3(), Nat.N1()).fill(0, 0, 0));
+			visionOutput.ifPresent((pose2dDoublePair) -> resetChassisPose(pose2dDoublePair.getFirst()));
+			poseEstimator.setVisionMeasurementStdDevs(new MatBuilder<>(Nat.N3(), Nat.N1()).fill(RobotMap.Vision.STANDARD_DEVIATION_VISION2D, RobotMap.Vision.STANDARD_DEVIATION_VISION2D, RobotMap.Vision.STANDARD_DEVIATION_VISION_ANGLE));
+		}
+		}
+
+
+	public boolean isAtPose(Pose2d goalPose) {
```

Comment:
If you need to document this function, you should refactor it to a diffrent function